vcr/vcr.go

```diff
@@ -224,8 +227,62 @@ func (c *vcr) Resolve(ID string) (did.VerifiableCredential, error) {
 	panic("implement me")
 }
 
-func (c *vcr) Verify(vc did.VerifiableCredential, credentialSubject interface{}, at time.Time) (bool, error) {
-	panic("implement me")
+func (c *vcr) Verify(vc did.VerifiableCredential, at time.Time) error {
+	// it must have valid content
+	validator, _ := credential.FindValidatorAndBuilder(vc)
+	if validator == nil {
+		return errors.New("unknown credential type")
+	}
+
+	if err := validator.Validate(vc); err != nil {
+		return err
+	}
+
+	// create correct challenge for verification
+	payload, err := generateCredentialChallenge(vc)
+	if err != nil {
+		return fmt.Errorf("cannot generate challenge: %w", err)
+	}
+
+	// extract proof, can't fail already done in generateCredentialChallenge
+	var proofs = make([]JSONWebSignature2020Proof, 0)
+	_ = vc.UnmarshalProofValue(&proofs)
+	proof := proofs[0]
+	splittedJws := strings.Split(proof.Jws, "..")
+	if len(splittedJws) != 2 {
+		return errors.New("invalid 'jws' value in proof")
+	}
+	sig, err := base64.RawURLEncoding.DecodeString(splittedJws[1])
+	if err != nil {
+		return err
+	}
+
+	// find key
+	pk, err := c.docResolver.ResolveSigningKey(proof.VerificationMethod.String(), &at)
+	if err != nil {
+		return err
+	}
+
+	// the proof must be correct
+	verifier, _ := jws.NewVerifier(jwa.ES256)
+	// the jws lib can't do this for us, so we concat hdr with payload for verification
+	challenge := fmt.Sprintf("%s.%s", splittedJws[0], base64.RawURLEncoding.EncodeToString(payload))
+	if err = verifier.Verify([]byte(challenge), sig, pk); err != nil {
+		return err
+	}
+
+	// next check timeRestrictions
+	if vc.IssuanceDate.After(at) {
```

Comment:
on or after, I'd think?

Reply:
I think it should be valid at the same nanosecond. So an inclusive start but exclusive end.